File: 34.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
class Solution {
  int lower_bound(vector<int> &nums, int l, int r, int target) {
    while (l <= r) {
      int m = (r - l) / 2 + l;
      if (nums[m] < target) {
        l = m + 1;
      } else {
        r = m - 1;
      }
    }
    return l;
  }

public:
  vector<int> searchRange(vector<int> &nums, int target) {
    int l = 0, r = nums.size() - 1;
    int start = lower_bound(nums, l, r, target);
    int end = lower_bound(nums, l, r, target + 1) - 1;
    if (start < nums.size() && nums[start] == target) {
      return {start, end};
    }
    return {-1, -1};
  }
};
int main() {
  vector<int> t1_in = {5, 7, 7, 8, 8, 10};
  cout << Solution().searchRange(t1_in, 8)[0];
  cout << " " << Solution().searchRange(t1_in, 8)[1];
}

File: 34.py
class Solution:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        if not nums:
            return [-1, -1]
        ans = [-1, -1]
        l, r = 0, len(nums) - 1
        # binary search 1, find left bound
        m = (r - l) // 2 + l
        while l <= r:
            m = (r - l) // 2 + l
            if target < nums[m]:
                r = m - 1
            elif target > nums[m]:
                l = m + 1
            # every case succeeding this one is when nums[m] == target:
            elif m == 0:
                break
            elif nums[m] != nums[m - 1]:
                break
            else:
                r = m - 1
        if nums[m] != target:
            return [-1, -1]
        ans[0] = m
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (r - l) // 2 + l
            if target < nums[m]:
                r = m - 1
            elif target > nums[m]:
                l = m + 1
            elif m == len(nums) - 1:
                break
            elif nums[m] != nums[m + 1]:
                break
            else:
                l = m + 1
        ans[1] = m
        return ans


print(Solution().searchRange([5, 7, 7, 8, 8, 10], 8))
print(Solution().searchRange([5, 7, 7, 8, 8, 10], 6))
